---
title: "Orchestrating Single-Cell Analysis with Bioconductor: Workshop Edition"
author: "Robert A. Amezquita, Stephanie C. Hicks"
date: "Last modified: May 27, 2019; Compiled; `r format(Sys.time(), '%B %d, %Y')`"
bibliography: "`r file.path(system.file(package='OSCABioc2019', 'vignettes'), 'bibliography.bib')`"
output:
  rmarkdown::html_document:
    highlight: pygments
    toc: true
    toc_depth: 3
    fig_width: 5
vignette: >
  %\VignetteIndexEntry{dummychapter1}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding[utf8]{inputenc}
---

```{r options, include=FALSE, message=FALSE, warning=FALSE}
knitr::opts_chunk$set(cache=FALSE, error=FALSE, message=FALSE, warning=FALSE)
```

# Orchestrating Single-Cell Analysis with Bioconductor: Workshop

## Instructors names and contact information

* Robert A. Amezquita, robert.amezquita@fredhutch.org
* Stephanie C. Hicks, shicks19@jhu.edu


## Workshop Description

This workshop gives an introductory overview of analyzing single-cell data, particularly RNA-seq, using Bioconductor software. This workshop will help participants to understand essential Bioconductor infrastructure, such as the *SingleCellExperiment* class, and various analytical routines using real-world data. Finally, this workshop is modeled after the manuscript *"Orchestrating Single-Cell Analysis with Bioconductor"* (Amezquita et al. 2019). Students will analyze provided example datasets on their personal laptop. This workshop will be a mixture of example code shown by instructors (available through this repository), short exercises, and discussion.


## Pre-requisites

* Basic knowledge of R syntax
* Some familiarity with S4 objects may be helpful, but is not required
* Some familiarity with tidyverse syntax and methods, such as the usage of pipes, may be helpful, but is not required
* Familiarity with high-throughput gene expression data as obtained from RNA-seq

Relevant background reading:

* [@OSCA]


## Workshop Participation

Students will be able to run code interactively during the workshop on their personal computers during a live demonstration of the code used herein, with ample opportunity for questions, answers, and discussion.


## _R_ / _Bioconductor_ packages used

* TBD: url here


## Time outline

1 hour workshop consisting of the following:

| Activity                              | Time |
|---------------------------------------|------|
| TBD                                   | TBDm |


## Workshop goals and objectives

### Learning goals

* Describe the framework needed to implement a basic single-cell RNA-seq analysis
* Describe the utility and design of the *SingleCellExperiment* object in the context of the analysis framework
* Identify critical steps within a typical single-cell RNA-seq analysis that can greatly influence end-results
* Understand the differences between single-cell vs. bulk based technologies, and the advantages/disadvantages of each


### Learning objectives

* Import single-cell RNA-seq data from raw counts into a *SingleCellExperiment* object
* Utilize the *SingleCellExperiment* object for annotation, subsetting, and processing via ad-hoc and established methods
* Produce descriptive plots from *SingleCellExperiment* objects to evaluate key quality control and processing steps

---

## Introduction

### The *SingleCellExperiment* Class

`r knitr::include_graphics(file.path(system.file(package='OSCABioc2019', 'vignettes'), "singlecellexperiment-overview.png"))`




